### ISSUE TITLE
Support complex numbers

### ISSUE THREAD

**Issue by PhilipVinc** [tillahoffmann/jax-mps#10]
Hey, very nice project!

This might be a tough one, because I suspect that metal does not support out of the box complex numbers, but hopefully I might be wrong...

And if they are not supported, it would be lovely if there was a way to 'rewrite' the HLO to support them (though this will be more complicated...).

**Comment by tillahoffmann** [comment]
I've added some basic complex number support in #11 although with several caveats. For example, gradients for holomorphic functions are currently untested and many complex-valued ops like FFTs are not yet implemented.

This will probably take a little while to implement—contributions welcome!

**Comment by tillahoffmann** [comment]
[`v0.9.3`](https://github.com/tillahoffmann/jax-mps/releases/tag/v0.9.3) (now on PyPI) adds some basic complex number support.

**Comment by joseph-long** [comment]
Great work! I use complex types in my models to implement Fourier / Fraunhofer diffraction, so FFTs would also be useful to me. It looks like you're making fast progress but I could also try taking an (AI-assisted) stab at it if you suggest a starting point. Just got the plugin built and ran the demo.

Not sure where else to ask this, but do you have a sense of the ceiling on speedups from MPS on Apple hardware? Is 3x the 'after a lot of optimization' number, or 'before much profiling' number?

**Comment by joseph-long** [comment]
After posting this I thought to check if FFTs had actually been implemented, looks like yes? https://github.com/tillahoffmann/jax-mps/blob/27471f2a18b38ceaa84c1a7e959a23b28d60b1ef/src/pjrt_plugin/ops/fft_ops.mm

```python
>>> x = jnp.arange(16).reshape(4,4).astype(jnp.complex64)
>>> jnp.fft.fft2(x)
Array([[120. +0.j,  -8. +8.j,  -8. +0.j,  -8. -8.j],
       [-32.+32.j,   0. +0.j,   0. +0.j,   0. +0.j],
       [-32. +0.j,   0. +0.j,   0. +0.j,   0. +0.j],
       [-32.-32.j,   0. +0.j,   0. +0.j,   0. +0.j]], dtype=complex64)
>>> x.device
MpsDevice(id=0)
```

Nice.

**Comment by tillahoffmann** [comment]
> After posting this I thought to check if FFTs had actually been implemented, looks like yes?

Yes, FFTs should work.

> I could also try taking an (AI-assisted) stab at it if you suggest a starting point.

That sounds great! If you're using Claude Code, it should automatically pick up the [`CLAUDE.md`](https://github.com/tillahoffmann/jax-mps/blob/main/CLAUDE.md) file. The [`.github/copilot-instructions.md`](https://github.com/tillahoffmann/jax-mps/blob/main/.github/copilot-instructions.md) also provides useful guidance for agents doing the coding so they know the criteria against which the work will be evaluated.

In terms of starting point, pick an operation you need and tell the agent to add test configs demonstrating that the ops fails. Git-stage the test files (so you know if the agent is cheating during implementation by modifying the tests). Then tell the agent to make the tests pass by implementing the op. Commit and push (make sure you have the [pre-commit hooks](https://github.com/tillahoffmann/jax-mps/blob/main/.pre-commit-config.yaml) installed).

> Not sure where else to ask this, but do you have a sense of the ceiling on speedups from MPS on Apple hardware? Is 3x the 'after a lot of optimization' number, or 'before much profiling' number?

Great question, and I unfortunately don't have a good answer. I think it will be very problem dependent, e.g., having large matrix multiplies could probably give significant speed-ups whereas smaller ops are likely much slower than on the CPU.

PyTorch reported a 5-10x speedup in training and 15-20x speedup in eval in their [blog post](https://pytorch.org/blog/introducing-accelerated-pytorch-training-on-mac/) for the initial MPS backend release. That's probably a good target. The blog post is from 2022 and the benchmark was run on an M1 chip; I'm not sure how much things have changed in terms of CPU/GPU relative performance.

**Comment by jkitchin** [comment]
Did you ever make progress on consolidating these two projects?

**Comment by tillahoffmann** [comment]
We've been tracking that over in #98. I'll follow up there and tag you for a progress update.
